### PR TITLE
Maid 1025: Test Pure Sentinel

### PR DESCRIPTION
**Pull request by mmoadeli**
The test passed for sentinel<1, 1> where one claim is followed by one key being sent.
 For sentinel<10, 10> sending 10 claims and, 10 claimant send public key of the 10 climants should provide 10X10 entries which should suffice for it to pass, while it does not.
To avoid code duplicates, a number of tests are presented in one test. They can be separated if it is required.

[Review on Reviewable](https://reviewable.io/reviews/maidsafe/sentinel/18)

### CODE REVIEW

**Comment by maidsafe-highfive** [comment]
Thanks for the pull request, and welcome! The MaidSafe team is excited to review your changes, and you should hear from @benjaminbollen (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way Github handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [CONTRIBUTOR.md](../blob/master/CONTRIBUTOR) for more information.

File: src/sentinel.rs [hunk 1]
```diff
@@ -208,4 +241,133 @@ mod test {
...
+        let signature = crypto::sign::sign_detached(&serialised_claim, &key_pair.1);
+        let climant_name = generate_random_name();
+
+        // first claim added should return AddResult::RequestKeys
+        match sentinel.add_claim(request.clone(), climant_name.clone(), signature.clone(), serialised_claim.clone()) {
+            Some(result) => match result {
+                AddResult::RequestKeys(source_name) => { assert_eq!(request.get_source(), source_name) },
+                AddResult::Resolved(_, _) => assert!(false)
+                },
+                None => assert!(true)
```
Comment (inetic) [inline src/sentinel.rs]: Could we change `assert!(true)` to just `_`?

Comment (mmoadeli) [inline src/sentinel.rs]: Peter, it does not compile if I do so

**Comment by benjaminbollen** [comment]
sorry I just noticed this

File: src/sentinel.rs (hunk 1 above)
Comment (inetic) [inline src/sentinel.rs]: It does, but it does nothing, so why have it there?

File: src/sentinel.rs [hunk 2]
```diff
@@ -197,10 +197,43 @@ impl<Request, Name>
...
+    }
+
+    fn generate_random_name() -> TestName {
+        let mut arr = [0u8;NAMESIZE];
+        for i in (0..NAMESIZE) { arr[i] = random::<u8>(); }
+        TestName { data : arr.to_vec() }
+    }
+
+    #[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
     struct TestRequest {
```
Comment (benjaminbollen) [inline src/sentinel.rs:222]: does TestRequest need more than just a simple usize ?  we can template Sentinel on Name = usize as well

File: src/sentinel.rs [hunk 3]
```diff
@@ -208,4 +241,133 @@ mod test {
...
+        // less than KEY_THRESHOLDS kyes received, should return None
+        for index in 0..KEY_THRESHOLDS - 1 {
+            match sentinel.add_keys(request.clone(), name_key_pairs[index].0.clone(), name_key_pairs.clone()) {
+                Some(_) => assert!(false),
+                None => assert!(true)
+            }
+        }
+
+        // KEY_THRESHOLDS kyes received, should not return none
+        match sentinel.add_keys(request.clone(), name_key_pairs[KEY_THRESHOLDS - 1].0.clone(), name_key_pairs.clone()) {
```
Comment (inetic) [inline src/sentinel.rs]: KEY_THRESHOLDS-1 is a constant, so `add_keys` will see all the keys as if they came from a single source. For the function to accept the keys, they need to come from KEY_THRESHOLDS of different sources.

File: src/sentinel.rs [hunk 4]
```diff
@@ -208,4 +241,133 @@ mod test {
         value : usize
```
Comment (benjaminbollen) [inline src/sentinel.rs:241]: can this not directly be a Vec < u8 >

File: src/sentinel.rs (hunk 3 above)
Comment (mmoadeli) [inline src/sentinel.rs]: the loop above that does so? Doesn't it?
this one is sent from the last guy

Comment (inetic) [inline src/sentinel.rs]: yes, you're right, sorry

File: src/sentinel.rs (hunk 1 above)
Comment (benjaminbollen) [inline src/sentinel.rs]: I agree with peter, please clean up the 'weird usage of asserts', I'll give an example below:

File: src/sentinel.rs [hunk 5]
```diff
@@ -208,4 +241,133 @@ mod test {
...
+        match sentinel.add_claim(request.clone(), climant_name.clone(), signature.clone(), serialised_claim.clone()) {
+            Some(result) => match result {
+                AddResult::RequestKeys(source_name) => { assert_eq!(request.get_source(), source_name) },
+                AddResult::Resolved(_, _) => assert!(false)
+                },
+                None => assert!(true)
+            }
+
+        // same claim added for the second time none to be returned
+        match sentinel.add_claim(request, climant_name, signature, serialised_claim) {
```
Comment (benjaminbollen) [inline src/sentinel.rs]: write this as assert!(sentinel.add_claim(request, climant_name, signature, serialised_claim).is_none());

Comment (mmoadeli) [inline src/sentinel.rs]: Ok

**Comment by mmoadeli** [comment]
the asserts are replaced.  thanks for the comments

File: src/sentinel.rs [hunk 6]
```diff
@@ -120,21 +120,21 @@ impl<Request, Name>
...
-            self.key_store.add_key(target, request.get_source(), public_key);
-        }
+          for (target, public_key) in keys {
+              self.key_store.add_key(target, sender.clone(), public_key);
+          }
 
-        self.claim_accumulator.get(&request)
-            .and_then(|(request, claims)| { self.resolve(request, claims) })
+          self.claim_accumulator.get(&request)
+              .and_then(|(request, claims)| { self.resolve(request, claims) })
```
Comment (inetic) [inline src/sentinel.rs]: Sorry for being a nazi, but why 6 spaces for indent?

Comment (mmoadeli) [inline src/sentinel.rs]: fixed .. was a copy / paste

**Comment by mmoadeli** [comment]
Now the test passes for Sentinel<10, 10> but passes for Sentinel<1, 1>

**Comment by mmoadeli** [comment]
It is ready for review

**Pull request merged**